Feature/handtering av ulik lonn for overtid

idatheo opened the pull request:
Adding 

- a separate database for alv economy to include salary data for employees. 
- controllers for adding salary data.
- method for calculating overtime salary at the given time an employee worked overtime

idatheo committed:
Legger til db economy for å lagre timeslønn for hver ansatt

idatheo committed:
Legger til tabell OvertimePayout i AlvEconomyData db for å lagre hvor mye en bruker får utbalt for overtid

idatheo committed:
Legger til loggikk for å beregne overtidsutbetiling på det tidspunktet du jobbet overtid, samt tester for dette

idatheo committed:
Legger til controller for å hente ut lønnshistorikk og registrere ny lønn for en ansatt

idatheo committed:
Renamer en metode for å registrere overtid i databasen

palbock commented on packages/api/AlvTime.Business/EconomyData/EmployeeSalaryDto.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class EmployeeSalaryDto
+    {
+        public int usiderId { get; set; }
```
Skal det være UserId?

idatheo replied:
gikk litt fort i svingene ja, fixed

palbock commented on packages/api/AlvTime.Business/EconomyData/IEmployeeHourlySalaryStorage.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public interface IEmployeeHourlySalaryStorage
+    {
+        decimal GetHouerlySalary(int userId, DateTime date);
```
Skriveleif: GetHourlySalary

idatheo replied:
Fixed

palbock commented on packages/api/AlvTime.Persistence/Repositories/AlvEconomyData/EmployeeHourlySalaryStorage.cs:
```diff
@@ -0,0 +1,92 @@
+        public decimal GetHouerlySalary(int userId, DateTime date)
+        {
+            var employeeWithSalaryData = _economyContext
+                    .EmployeeHourlySalaries.Where(hs => hs.UserId == userId)
```
Sleng på en .ToList() på slutten her så queryet faktisk kjører, hvis ikke blir du stuck med en IQueryable :)

idatheo replied:
Fixed

palbock commented on packages/api/AlvTime.Persistence/Repositories/AlvEconomyData/EmployeeHourlySalaryStorage.cs:
```diff
@@ -0,0 +1,92 @@
+        public List<EmployeeSalaryDto> GetEmployeeSalaryData(int userId)
+        {
+            var employeeWithSalaryData = _economyContext
+                    .EmployeeHourlySalaries.Where(hs => hs.UserId == userId)
```
.ToList() på slutten her også

idatheo replied:
Fixed

palbock commented on packages/api/AlvTime.Persistence/Repositories/AlvEconomyData/OvertimePayoutStorage.cs:
```diff
@@ -0,0 +1,117 @@
+        public OvertimePayoutResponsDto DeleteOvertimePayout(int userId, DateTime date)
+        {
+            var overtimePayout =
+                _economyContext.OvertimePayouts.FirstOrDefault(op => op.UserId == userId && op.Date.Equals(date));
```
Lurer på om det kanskje finnes en bedre måte å matche Payout på enn å sjekke dato. Ser for meg at det kan bli litt krøll med tidspunkt der. Er det mulig å innføre en slags Id i OvertimePayouts som matche Id til PaidOvertime i den andre contexten? Alternativt kunne man flyttet hele PaidOvertime tabellen fra Alvtime_dbContext til AlvEconomyDataContext, men det blir kanskje mye styr...

idatheo replied:
Jeg er enig, skal sjekke. (Mulig jeg kan bruke Id'en fra paid overtime?)

idatheo replied:
Byttet det ut med Iden fra PaidOvertime. Da skal det i teorien bli helt riktig

palbock commented on packages/api/AlvTime.Persistence/Repositories/AlvEconomyData/OvertimePayoutStorage.cs:
```diff
@@ -0,0 +1,117 @@
+
+            return salary;
+        }
+        private List<OvertimeEntry> GetOvertimeEntriesForPayout(List<OvertimeEntry> overtimeEntries, decimal hoursForPayout)
```
Klarer ikke helt å følge denne metoden hele veien gjennom. Kanskje du kan forklare meg den på Alvtime-torsdag? :)

idatheo replied:
Ja, veldig gjerne, for det burde være en bedre/ penere måte å gjøre dette på, men var den eneste jeg kom på som faktisk ga riktige resultater

palbock commented on packages/api/AlvTime.Persistence/Repositories/AlvEconomyData/EmployeeHourlySalaryStorage.cs:
```diff
@@ -0,0 +1,92 @@
+            return new()
+            {
+                usiderId = employeeHourlySalary.UserId, FromDate = employeeHourlySalary.FromDateInclusive,
+                ToDate = employeeHourlySalary.ToDate, HourlySalary = employeeHourlySalary.HourlySalary
```
Vi har ikke vært helt konsekvent på det, men det kan være penere (for frontend) å returnere en dato som string som ikke inneholder tidspunkt. Altså kjøre .ToString("yyyy-MM-dd", CultureInfo.InvariantCulture) etter datoen.

idatheo replied:
Ah, okey. Jeg bruker samme objektet som input i POST EmployeeSalary. Sender frontend da inn dato som string også? Burde jeg opprette eget objekt for input og eget for retur? (Kanskje se litt på den også sammen?)

palbock replied:
Frontend sender dato inn som dato. Enesten fordelen med å returnere det ut til frontend som string er at da slipper frontend å gjøre noe logikk for å formatere det og kan bare presentere responsen as is. Kan godt hende det er en gode ide å lage to objekter. Vi kan godt se på det sammen i morgen.

idatheo replied:
Okey, jeg laget det om til to ulike records (en for request og en for respons) som vi diskuterte, meeeen jeg gjordemuligens noe rart. I request, trenger jeg på en tidspunkt ID'en, men jeg vil ikke åpne for å ta det som input når du skal registrere ny lønn, og jeg kan ikke bruke responsobjektet fordi det har dato som string. Jeg innførte derfor en tredje record som har med seg Id i tillegg til info fra requesten. Kan du se på det om det ble veldig forvirrende?

![image](https://user-images.githubusercontent.com/79319625/130633050-120335e0-acf8-4517-982f-ae1545233c10.png)

idatheo replied:
Ref kommentar på EmployeeSalary: renamer til EmployeeSalaryDto og fjernet arv da dette var økte lesbarheten

palbock commented on packages/api/AlvTimeWebApi/Controllers/EconomyData/SalaryController.cs:
```diff
@@ -0,0 +1,34 @@
+        [HttpPost("/EmployeeSalary")]
+        public ActionResult RegisterHourlySalary([FromBody] EmployeeSalaryDto employeeSalaryData)
+        {
+            _employeeHourlySalaryStorage.RegisterHourlySalary(employeeSalaryData);
```
Vet vi stort sett har gjort det på denne måten ved å gå rett fra controller til storage/repository, men det er et ønske om å skrive om til clean architecture. Dvs at controller sender requesten til en service som ligger i business-laget, som så kaller på persistence-laget gjennom et interface. På den måten risikerer vi ikke at noe database-logikk lekker ut til de andre lagene og vi unngår så mye businesslogikk i database-laget. Jeg har startet på en refaktorering i branchen som heter "refactor_to_onion_architecture", der kan du se på AccessTokenController og følge flyten gjennom der hvis du har lyst til å gjøre det på denne måten.

idatheo replied:
Fixed-ish på denne PRen. Introdusert SalaryService

palbock commented on packages/api/AlvTime.Business/EconomyData/OvertimePayoutResponsDto.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace AlvTime.Business.EconomyData
+{
+    public class OvertimePayoutResponsDto
```
Absolutt ikke nødvendig og kanskje mer smak og behag enn noe annet, men nå som vi bruker .NET 5 går det an å bruke records som er perfekte til Dto'er som bare inneholder data som sendes frem og tilbake og ikke skal endres. https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record

idatheo replied:
Byttet til records :)

palbock commented on packages/api/AlvTimeWebApi/Controllers/EconomyData/SalaryController.cs:
```diff
@@ -0,0 +1,34 @@
+
+        [AuthorizeAdmin]
+        [HttpPost("/EmployeeSalary")]
+        public ActionResult RegisterHourlySalary([FromBody] EmployeeSalaryDto employeeSalaryData)
```
Burde denne returnere noe?

idatheo replied:
Returnerer nå ny, registrert lønn: 

return Ok(salary);

idatheo committed:
småfikser fra code review

idatheo committed:
Flytter en del av logikken fra EmployeeHourlySalaryStorage og OvertimePayoutStorage ut i SalaryService

idatheo committed:
Skriver om metoden RegisterOvertimePayout, og flytter metoden over i FlexhourStorage

idatheo committed:
Legger til Id fra PaidOvertime (fra AlvDevDb) i OvertimePayout for å gjøre oppslag for å slette evt. overtidsbetaling mer korrekt

idatheo committed:
Endrer på DeleteOvertimePayout til å bruke RegisteredPaidOvertimeId for å unngå problem ved sletting av overtid.
Endrer på OvertimePayoutResponsDto til å ha dato som string for evt. bruk av frontend

idatheo committed:
Fjerner ubrukte usings i diverse filer

idatheo committed:
Fjerner ubrukte metoder for å hente ut timeslønn. Endrer på SaveOvertimePayout til å returnere OvertimePayoutRespons. Småtweaking i tester

idatheo committed:
Endrer OvertimePayoutResponsDto, RegisterOvertimePayoutDto og EmployeeSalaryDto til records, og fjerner Dto post-fix

idatheo committed:
Endrer slik at RegisterHourlySalary returnerer ny lønn

idatheo committed:
Endrer til at EmployeeSalaray returnerer dato som string til frontend

idatheo committed:
Merge branch 'master' into feature/handtering-av-ulik-lonn-for-overtid

palbock commented on packages/api/AlvTime.Business/EconomyData/EmployeeSalaryRequest.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public record EmployeeSalaryRequest
+    {
+        public int UsiderId { get; init; }
```
Skal det stå "UserId"? :P

idatheo replied:
Fixed

palbock commented on packages/api/AlvTime.Business/EconomyData/EmployeeSalaryRespons.cs:
```diff
@@ -0,0 +1,13 @@
+    public record EmployeeSalaryRespons 
+    {
+        public int Id { get; init; }
+        public int UsiderId { get; init; }
```
UserId?

idatheo replied:
Fixed

palbock commented on packages/api/AlvTime.Business/EconomyData/EmployeeSalaryRequest.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace AlvTime.Business.EconomyData
+{
+    public record EmployeeSalaryRequest
```
Records kan også skrive på denne måten: `public record EmployeeSalaryRequest (int UserId, decimal HourlySalary, DateTime FromDate, DateTime? ToDate)` og det vil være helt likt som det du har skrevet nå, hvis du vil ha det litt mer konsist :)

idatheo replied:
Fixed! (Pluss et par følgefeil)

palbock commented on packages/api/AlvTime.Persistence/Repositories/FlexhourStorage.cs:
```diff
@@ -399,12 +400,45 @@ private decimal GetHoursAfterCompRate(List<OvertimeEntry> overtimeEntries, decim
 
         return totalPayout;
     }
+    public decimal RegisterOvertimePayout(List<OvertimeEntry> overtimeEntries, int userId, GenericHourEntry requestedPayout, int paidOvertimeId)
```
Det kunne vært digg å få flytta den logikken her ut til SalaryService, men tenker uansett å flytte all logikken ut fra FlexhourStorage på sikt så ikke noe stress om du ikke gjør det nå.

idatheo replied:
Fikk ikke flyttet den ut uten å lage en sirkulær dependency (?) mellom persistance- og businessprosjektet, den må flyttes til salary service når vi går over til services tenker jeg

idatheo committed:
Småfikser etter codereview. Oppretter script for å opprette AlvEconomyDatabase samt legger inn data i EmployeeHourlySalaries for AhreKetil & co.

idatheo committed:
Merge branch 'master' into feature/handtering-av-ulik-lonn-for-overtid

idatheo committed:
Legger til noen tester for beregning av lønn ved utbetalt overtid, og fikser på koden slik at testene kjører grønt

idatheo committed:
Fiks på feil i testene

idatheo committed:
Adding more tests for overtimepayout calculations

idatheo committed:
Legger til endringer for å støtte avspasering mtp ulik comprate

idatheo committed:
Fiks på at noen tester feilet pga man måtte gruppere overtimeEntries på kompensasjonsrate før man sorterte på dato, i stedet for å gruppere på lønn. (Fungerende kode med dårlige variabelnavn, neste committ fikser variabelnavn)

idatheo committed:
Merge branch 'master' into feature/handtering-av-ulik-lonn-for-overtid

idatheo committed:
Legger til en test pga endringer i gruppering på comp rate og sortering på lønn. Endrer på variabel og metodenavn for å øke lesbarhet

idatheo committed:
Merge branch 'master' into feature/handtering-av-ulik-lonn-for-overtid

idatheo committed:
Flytter felles brukte metoder i FlexiHour testene ut i en FlexiHoursTestUtils. Flytter tester for RegisterOvertimePayout (flexhourstarage) ut i egen separat fil. Legger inn "formlene" for utregning i Assertene i RegisterPaidOVertimePayoutTests for at de skal være lettere å forstå/ vedlikeholde

idatheo committed:
REnamer metode for å filtrere ut avspasering og tidligere utbetalinger (Flexhour storage)

palbock commented on packages/api/AlvTime.Business/EconomyData/EmployeeSalary.cs:
```diff
@@ -0,0 +1,9 @@
+namespace AlvTime.Business.EconomyData
+{
+    public record EmployeeSalary(int UserId, decimal HourlySalary, DateTime FromDate, DateTime? ToDate, int Id) : 
+        EmployeeSalaryRequest(UserId, HourlySalary, FromDate, ToDate)
```
Trenger EmpoyeeSalary å arve fra EmployeeSalaryRequest eller blir det bare mer forvirrende?

idatheo replied:
Nei, egentlig burde det jo være omvendt, og så er jeg litt lite glad i å ikke koble sammen objekter som åpenbart har noe med hverandre å gjøre. Fjernet arv, for jeg er enig :)

palbock commented on packages/api/AlvTime.Business/EconomyData/EmployeeSalary.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace AlvTime.Business.EconomyData
+{
+    public record EmployeeSalary(int UserId, decimal HourlySalary, DateTime FromDate, DateTime? ToDate, int Id) : 
```
Kan den hete EmployeeSalaryDto istedet, så blir det litt mer ulikt alle de andre "EmployeeSalary" klassene? :P

idatheo replied:
Fixed 👍

palbock commented on packages/api/AlvTime.Business/EconomyData/OvertimePayoutRespons.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace AlvTime.Business.EconomyData
+{
+    public record OvertimePayoutRespons (int Id, int UserId, string Date, decimal TotalPayout, int PaidOvertimeId)
```
Response med `e` på slutten :)

idatheo replied:
Jeg er ikke vant til å kode på norsk i det hele tatt nei... 🙈

palbock commented on packages/api/AlvTime.Persistence/Repositories/AlvEconomyData/OvertimePayoutStorage.cs:
```diff
@@ -0,0 +1,62 @@
+            return new OvertimePayoutRespons(
+                overtimePayoutEntity.Id, 
+                overtimePayoutEntity.UserId,
+                overtimePayoutEntity.Date.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture),
```
Kanskje denne dato mappingen burde skje i API'et siden det kun er frontend som bryr seg om det?

idatheo replied:
Hehehehehe, denne når ikke frontend heller for øyeblikket, så ehm ja. Bytter til DateTime

palbock commented on packages/api/AlvTime.Business/EconomyData/EmployeeSalaryRequest.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace AlvTime.Business.EconomyData
+{
+    public record EmployeeSalaryRequest (int UserId, decimal HourlySalary,  DateTime FromDate,DateTime? ToDate)
```
Hadde det vært mulig å fjerne ToDate fra requesten her, og bare gjøre slik at den blir satt automatisk når det kommer en ny lønn etterpå? Det gjør du vel allerede når en ny lønn registreres, så kanskje det er unødvendig at man spesifiserer en "til dato". Er vel sjeldent man vet akkurat hvilken til dato man vil ha og

idatheo replied:
Ja, du har et poeng der. Fjerner ToDate. Det slo meg nå at ToDate må settes når en ansatt slutter, det har jeg glemt

idatheo replied:
Oppretter egen issue på folk som slutter. På tide å få denne ut i prod 😅

palbock commented on packages/api/Tests/UnitTests/Flexihours/RegisterPaidOvertimePayoutTests.cs:
```diff
@@ -0,0 +1,859 @@
+            {
+                UserId = userId,
+                HourlySalary = 100.0M,
+                FromDateInclusive = new DateTime(day: 01, month: 07, year: 2019),
```
Hvorfor har FromDateInclusive og ToDate forskjellig format på DateTime'n sin? 😄

idatheo replied:
copy paste feil. Fixed

palbock commented on packages/api/AlvTime.Persistence/Repositories/FlexhourStorage.cs:
```diff
@@ -329,27 +323,31 @@ public PaidOvertimeEntry RegisterPaidOvertime(GenericHourEntry request, int user
+            _context.PaidOvertime.Remove(paidOvertime);
             _context.SaveChanges();
 
+            _salaryService.DeleteOvertimePayout(userId, paidOvertime.Id);
```
Hva skjer her dersom `_context.PaidOvertime.Remove` går gjennom, men `_salaryService.DeleteOvertimePayout` feiler? Da tror jeg det blir noe trøbbel. Kan vi vurdere å wrappe dette i en transaksjon? Se for eksempel svaret på denne posten: https://stackoverflow.com/questions/815586/using-transactions-or-savechangesfalse-and-acceptallchanges

idatheo replied:
Må legge både registrering og sletting i transaksjon

idatheo replied:
Fixed (lagt inn transaction på både registrering og sletting)

palbock commented on packages/api/AlvTime.Persistence/Repositories/FlexhourStorage.cs:
```diff
@@ -329,27 +323,31 @@ public PaidOvertimeEntry RegisterPaidOvertime(GenericHourEntry request, int user
             _context.SaveChanges();
 
+            _salaryService.DeleteOvertimePayout(userId, paidOvertime.Id);
+
```
Generelt spørsmål: hva er konsekvensen av å ha foreign keys på tvers av databaser? Altså for eksempel at UserId i økonomidatabasen peker på en bruker i den "vanlige" databasen. Kan vi havne i en situasjon hvor dette kan skape problemer? Aner ikke selv, kanskje bare noe å ha i bakhodet 😃

idatheo replied:
Usikker på konsekvenser. Det er jo ingen direkte kobling :/ Tror kanskje vi bare må holde et øye med det?

palbock commented on packages/api/AlvTimeWebApi/Controllers/EconomyData/SalaryController.cs:
```diff
@@ -0,0 +1,49 @@
+                employeeHourlySalary.HourlySalary, 
+                employeeHourlySalary.FromDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture),
+                employeeHourlySalary.ToDate.HasValue
+                    ? employeeHourlySalary.ToDate.Value.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture)
```
Det finnes en klasse som heter `DateFormatter` som kan brukes til å gjøre denne datomappingen

idatheo replied:
Fixed

palbock commented on packages/api/AlvTimeWebApi/Controllers/EconomyData/SalaryController.cs:
```diff
@@ -0,0 +1,49 @@
+
+        [AuthorizeAdmin]
+        [HttpPost("/EmployeeSalary")]
+        public ActionResult RegisterHourlySalary([FromBody] EmployeeSalaryRequest employeeSalaryData)
```
Spesifisere return type

idatheo replied:
Fixed

palbock commented on packages/api/AlvTime.Business/Services/SalaryService.cs:
```diff
@@ -0,0 +1,47 @@
+            return _employeeHourlySalaryStorage.GetEmployeeSalaryData(userId);
+        }
+
+        public EmployeeSalaryRequest ToEmployeeSalary(EmployeeSalaryRequest employeeSalaryRequest)
```
Vi vil slippe hele denne mappingen dersom vi dropper ToDate fra EmployeeSalaryRequest

idatheo replied:
Fjernet

palbock commented on packages/api/AlvTimeWebApi/Controllers/EconomyData/SalaryController.cs:
```diff
@@ -0,0 +1,49 @@
+
+        [AuthorizeAdmin]
+        [HttpGet("/EmployeeSalary")]
+        public ActionResult<List<EmployeeSalaryRespons>> GetEmployeeSalaryData(int userId)
```
Kunne det vært en idè å la denne metoden ta inn en liste med bruker id'er så man kan hente lønnsinfo om flere ansatte samtidig?

idatheo replied:
Fixed

palbock commented on packages/api/AlvTimeWebApi/Controllers/EconomyData/SalaryController.cs:
```diff
@@ -0,0 +1,49 @@
+{
+    [Route("api/EconomyData")]
+    [ApiController]
+    public class SalaryController : Controller
```
Burde vi hatt en controller for å hente ut utbetalingsinfo også? Altså antall kroner. Kanskje kun gjøre den tilgjengelig for admin foreløpig, men på sikt ser jeg for meg at alle ansatte skal kunne se utbetalingene sine i kroner.

idatheo replied:
Hvilken av utbetalingene tenker du da på? Vanlig lønnslipp eller overtidsutbetalingen? (For hvor mye man får overtidsbetalt er ønsket av flere allerede, tenkte jeg kunne slå meg løs på det etter denne issuen)

idatheo replied:
Oppretter egen issue på dette

palbock commented on packages/api/AlvTime.Persistence/Repositories/FlexhourStorage.cs:
```diff
@@ -399,12 +397,161 @@ private decimal GetHoursAfterCompRate(List<OvertimeEntry> overtimeEntries, decim
+
+    public decimal RegisterOvertimePayout(List<OvertimeEntry> overtimeEntries, int userId, GenericHourEntry requestedPayout, int paidOvertimeId)
+    {
+        var salaryData = _salaryService.GetEmployeeSalaryData(userId).OrderBy(x => x.FromDate).ToList();
```
Vi kan sikkert refaktorere dette senere, men det ser litt funky ut at flyten går fra service -> repository -> service -> repository. Optimalt burde det kun vært èn gang innom hvert sted. Men bare la det være sånn for nå, hele FlexhourStorage er jo bare megaspagetti så det må gjøres noe med på en litt større skala 😄

idatheo replied:
Ja, jeg er enig i at dette ikke er optimalt. Det ble litt fusjon av pre-onion architectur og onion architecture. Tenker at jeg kan refakturere litt når det kommer

palbock commented on packages/api/AlvTime.Persistence/Repositories/FlexhourStorage.cs:
```diff
@@ -399,12 +397,161 @@ private decimal GetHoursAfterCompRate(List<OvertimeEntry> overtimeEntries, decim
+            var overtimeEntriesForGivenCompRate = new List<OvertimeEntryWithSalary>();
+            foreach (var overtimeEntry in overtimeEntryGroupForGivenCompRate)
+            {
+                var salaryForOvertimeEntry = salaryData.FirstOrDefault(x => overtimeEntry.Date >= x.FromDate && (x.ToDate == null || overtimeEntry.Date < x.ToDate)).HourlySalary;
```
Gjerne bruk et annet variabelnavn enn `x` :)

idatheo replied:
wops

palbock commented on packages/api/AlvTime.Persistence/Repositories/FlexhourStorage.cs:
```diff
@@ -399,12 +397,161 @@ private decimal GetHoursAfterCompRate(List<OvertimeEntry> overtimeEntries, decim
+            }
+        }
+        
+        indexForCompRateList = indexForCompRateList.OrderBy(x => x.compRate).ToList();
```
Tror det holder å bare kjøre `indexForCompRateList.OrderBy(x => x.compRate).ToList();`, trenger ikke assigne den på nytt. Dvs droppe `=`

idatheo replied:
Jeg fik faktisk ikke til det. Listen oppdaterte seg ikke :/ så måtte kjøre denne litt uortodokse varianten

idatheo replied:
Burde jeg da kanskje heller gitt listen et nytt navn/ opprettet ny liste? Av typen: `var orderedIndexList = indexForCompRateList.OrderBy(x => x.compRate).ToList(); `

idatheo replied:
N/A

palbock commented on packages/api/Tests/UnitTests/Flexihours/RegisterPaidOvertimePayoutTests.cs:
```diff
@@ -0,0 +1,859 @@
+            _context.Hours.Add(flexTimeEntry);
+            _context.CompensationRate.Add(FlexiHoursTestUtils.CreateCompensationRate(flexTask.Id, 1.0M));
+
+            var expectedPayout1 = salary.HourlySalary*(0.5M*5+1.0M*1M);
```
Føler testen blir lettere å lese om expectedPayout ligger under utbetalingsbestillingene, men det er kanskje bare meg. Opp til deg 😃

idatheo replied:
Jeg er enig egentlig. Maks overtenking

idatheo committed:
Småfikser basert på tilbakemelding fra Code review. Litt renaming og flytting av variabel deklarasjon for bedre lesbarhet

idatheo committed:
Endrer gruperingen på comprate for overtidsberegning fra å bruke matrise med List<List...>> til dictionary med comp rate som key

idatheo committed:
Fjerner ToDate på registrering av ny Lønn

idatheo committed:
fix inconsistent initialization of new DateTime(...) in SalaryServiceTests, RegisterPaidOvertimePayoutTests and REgisterOvertimePayoutTests

idatheo committed:
Tar i bruk DAteFormatter i SalaryController

idatheo committed:
Endrer på endepunkt for å hente ut lønnsdata om en ansatt til å kunne hente ut lønnsdata for flere ansatte i SalaryController

idatheo committed:
Legger til bruk av transaction ved registrering og sletting av overtidsutbetaling

idatheo committed:
Merge branch 'master' into feature/handtering-av-ulik-lonn-for-overtid

FredrikAugust commented:
Denne branchen og PRen er ganske _stale_. Kan den lukkes eller skal den jobbes videre med, @palbock90 og @idatheo
